sys/net/routing/ng_rpl/ng_rpl_control_messages.c

```diff
@@ -0,0 +1,603 @@
+        dodag_conf->length = NG_RPL_OPT_DODAG_CONF_LEN;
+        dodag_conf->flags_a_pcs = 0;
+        dodag_conf->dio_int_doubl = dodag->dio_interval_doubl;
+        dodag_conf->dio_int_min = dodag->dio_min;
+        dodag_conf->dio_redun = dodag->dio_redun;
+        dodag_conf->max_rank_inc = byteorder_htons(dodag->instance->max_rank_inc);
+        dodag_conf->min_hop_rank_inc = byteorder_htons(dodag->instance->min_hop_rank_inc);
+        dodag_conf->ocp = byteorder_htons(dodag->instance->of->ocp);
+        dodag_conf->reserved = 0;
+        dodag_conf->default_lifetime = dodag->default_lifetime;
+        dodag_conf->lifetime_unit = byteorder_htons(dodag->lifetime_unit);
+        pos += sizeof(*dodag_conf);
+    }
+
+    dodag->dodag_conf_counter++;
+    _ng_rpl_send(pkt, NULL, destination, &dodag->dodag_id);
+}
+
+void ng_rpl_send_DIS(ng_rpl_dodag_t *dodag, ng_ipv6_addr_t *destination)
+{
+    (void) dodag;
+    ng_pktsnip_t *pkt;
+    ng_icmpv6_hdr_t *icmp;
+    ng_rpl_dis_t *dis;
+    /* TODO: Currently the DIS is too small so that wireshark complains about an incorrect
+     * ethernet frame check sequence. In order to prevent this, 4 PAD1 options are added.
+     * This will be addressed in follow-up PRs */
+    int size = sizeof(ng_icmpv6_hdr_t) + sizeof(ng_rpl_dis_t) + 4;
+
+    if ((pkt = ng_icmpv6_build(NULL, NG_ICMPV6_RPL_CTRL, NG_RPL_ICMPV6_CODE_DIS, size)) == NULL) {
+        DEBUG("RPL: Send DIS - no space left in packet buffer\n");
+        return;
+    }
+
+    icmp = (ng_icmpv6_hdr_t *)pkt->data;
+    dis = (ng_rpl_dis_t *)(icmp + 1);
+    dis->flags = 0;
+    dis->reserved = 0;
+    /* TODO: see above TODO */
+    memset((dis + 1), 0, 4);
+
+    _ng_rpl_send(pkt, NULL, destination, (dodag ? &dodag->dodag_id : NULL));
+}
+
+void ng_rpl_recv_DIS(ng_rpl_dis_t *dis, ng_ipv6_addr_t *src, ng_ipv6_addr_t *dst, uint16_t len)
+{
+    /* TODO handle Solicited Information Option */
+    (void) dis;
+    (void) len;
+
+    if (ng_ipv6_addr_is_multicast(dst)) {
+        for (uint8_t i = 0; i < NG_RPL_DODAGS_NUMOF; ++i) {
+            if (ng_rpl_dodags[i].state != 0) {
+                trickle_reset_timer(&ng_rpl_dodags[i].trickle);
+            }
+        }
+    }
+    else {
+        for (uint8_t i = 0; i < NG_RPL_DODAGS_NUMOF; ++i) {
+            if (ng_rpl_dodags[i].state != 0) {
+                ng_rpl_dodags[i].dodag_conf_counter = 0;
+                ng_rpl_send_DIO(&ng_rpl_dodags[i], src);
+            }
+        }
+    }
+}
+
+bool _parse_options(ng_rpl_dodag_t *dodag, ng_rpl_opt_t *opt, uint16_t len, ng_ipv6_addr_t *src)
+{
+    uint16_t l = 0;
+    ng_rpl_opt_target_t *first_target = NULL;
+    while(l < len) {
+        if ((opt->type != NG_RPL_OPT_PAD1) && (len < opt->length + sizeof(ng_rpl_opt_t) + l)) {
+            /* return false to delete the dodag,
+             * because former options may also contain errors */
+            return false;
+        }
+
+        switch(opt->type) {
```

Comment:
option types should be checked against the parsed packet type to discard for example DIO packets which contain DAO options etc.